src/ConnectedMode.UnitTests/BranchMatcherTests.cs

```diff
@@ -0,0 +1,211 @@
+            var devBranchCommit = new CommitWrapper(2);
+            var headBranchCommit = new CommitWrapper(3);
+
+            var masterBranch = CreateBranch("master", masterCommit);
+            var devBranch = CreateBranch("dev", devBranchCommit, masterCommit);
+            var headBranch = CreateBranch("branch", headBranchCommit, devBranchCommit, masterCommit);
+
+            var repo = CreateRepo(headBranch, masterBranch, devBranch);
+
+            var testSubject = new BranchMatcher(service, repo);
+
+            var result = await testSubject.GetMatchedBranch("projectKey");
+
+            result.Should().Be("dev");
+        }
+
+        /*
+         * Master
+         * |->Commit1
+         * |
+         * |->Commit2
+         * |
+         * |->Commit3
+         * |    |
+         * |    |---->Branch1
+         * |    |       |
+         * |    |       |->Commit4
+         * |    |       |
+         * |    |       |->Commit5
+         * |    |       |
+         * |    |
+         * |    |---->Branch2 (closest branch)
+         * |    |       |
+         * |    |       |->Commit6
+         * |    |       |
+         * |    |
+         * |    |---->Branch3 (Head)
+         * |    |       |
+         * |    |       |->Commit7
+         * |    |       |
+         * |
+         * |->Commit8
+         * |
+         * |->Commit9
+         * |
+         * |->Commit10
+         * |
+         */
+        [TestMethod]
+        public async Task GetMatchedBranch_MultipleCandidate_ChooseClosestMatch()
+        {
+            var service = CreateService("master", "Branch1", "Branch2");
+
+            var commit1 = new CommitWrapper(1);
+            var commit2 = new CommitWrapper(2);
+            var commit3 = new CommitWrapper(3);
+            var commit4 = new CommitWrapper(4);
+            var commit5 = new CommitWrapper(5);
+            var commit6 = new CommitWrapper(6);
+            var commit7 = new CommitWrapper(7);
+            var commit8 = new CommitWrapper(8);
+            var commit9 = new CommitWrapper(9);
+            var commit10 = new CommitWrapper(10);
+            
+            var masterBranch = CreateBranch("master", commit10, commit9, commit8, commit3, commit2, commit1);
+            var branch1 = CreateBranch("branch1", commit5, commit4, commit3, commit2, commit1);
+            var branch2 = CreateBranch("branch2", commit6, commit3, commit2, commit1);
+            var branch3 = CreateBranch("branch3", commit7, commit3, commit2, commit1);
+
+            var repo = CreateRepo(branch3, masterBranch, branch1, branch2);
+
+            var testSubject = new BranchMatcher(service, repo);
+
+            var result = await testSubject.GetMatchedBranch("projectKey");
+
+            result.Should().Be("branch2");
+        }
+
+        [TestMethod]
+        public async Task GetMatchedBranch_NoMatchingBranch_ChooseMain()
+        {
+            var service = CreateService("master", "branch1", "branch2");
```

Comment:
Naming: `service` is very general. `sonarService` / `CreateSonarService` would give a bit more context.